File: api/nim/pl0t.nim
import std/httpclient, std/strformat, std/os, std/tables, std/strutils

when (compiles do: import base/[basem, jsonm]):
  import base/[basem, jsonm]
else:
  import std/json, std/jsonutils
  proc to_json_hook*(n: JsonNode): JsonNode = n
  template throw(message: string) = raise Exception.new_exception(message)

template convert_to_json(data: JsonNode, pretty: bool): string =
  when compiles(data.to_s pretty):
    data.to_s pretty
  else:
    if pretty: data.pretty else: $data


# publish, unpublish -------------------------------------------------------------------------------
var plot_api_token* = get_env("plot_api_token", "")

proc build_plot_url(url: string): string =
  if plot_api_token == "":
    throw "plot_api_token not defined, define it explicitly or as as `plot_api_token` environment variable"
  fmt"{url}?api_token={plot_api_token}"

proc publish*(data: JsonNode, url: string, pretty = true): void =
  let url   = build_plot_url url
  let client = new_http_client()
  defer: client.close
  let json = convert_to_json(data, pretty)
  discard client.post_content(url, json)

proc unpublish*(url: string): void =
  let url = build_plot_url url
  let client = new_http_client()
  defer: client.close
  discard client.delete(url)


# Standalone HTML template -------------------------------------------------------------------------
# standalone begin
let standalone = """<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width,initial-scale=1">

  <title>pl0t</title>

  <link rel="icon" type="image/png" href="http://files.pl0t.com/view-1/favicon.ico">

  <!-- PL0T begin, putting it at the end of the page to avoid blocking other content -->
  <script>
    window.env = {
      base_url: "http://files.pl0t.com/view-1"
    }
  </script>
  <link rel="stylesheet" href="http://files.pl0t.com/view-1/releases/2021-08-30-b1cf3c/bundle.css">
  <script defer src="http://files.pl0t.com/view-1/releases/2021-08-30-b1cf3c/bundle.js"></script>
  <!-- PL0T end -->
</head>
<body>

<!-- PL0T data, type could be 'yaml', 'json' or 'md' -->
<script id="data" type="{type}">
{data}
</script>


<script>
  window.on_plot_loaded = function() {
    window.plot_api.run()
  }
</script>

</body>
</html>
"""
# standalone end


# Page ---------------------------------------------------------------------------------------------
type Page* = ref object
  attrs: JsonNode
  page*: seq[JsonNode]

proc init*(_: type[Page], attrs: JsonNode): Page =
  Page(attrs: attrs)

proc text*(page: var Page, id: string, text: string): void =
  page.page.add (id: id, text: text).to_json

proc table*[D](page: var Page, id: string, data: D, table: JsonNode = newJObject()): void =
  page.page.add (id: id, table: table, data: data).to_json

proc chart*[D](page: var Page, id: string, data: D, chart: JsonNode): void =
  page.page.add (id: id, chart: chart, data: data).to_json

proc image*(page: var Page, id: string, data: JsonNode): void =
  page.page.add (id: id, image: data).to_json

proc as_json(page: Page): JsonNode =
  var json = ($(page.attrs)).parse_json # Have no idea how to clone it otherwise
  json["page"] = page.page.to_json
  json

proc publish*(page: Page, url: string, pretty = true): void =
  page.as_json.publish(url, pretty)

proc save*(page: Page, path: string, pretty = true): void =
  let json = page.as_json
  let json_s = convert_to_json(json, pretty)
  let html = standalone.replace("{type}", "json").replace("{data}", json_s)
  if file_exists path: remove_file path
  write_file(path, html)

File: api/ruby/pl0t.rb
class Page
  @app = ['page', 0.1]

  attr_accessor :id, :title, :desc, :css, :page

  def initialize params = {}
    self.page = []
    params.each do |k, v|
      send("#{k}=", v)
    end
  end

  def text id, text
    page << { id: id, text: text }
  end

  def table id, data, table
    page << { id: id, table: table, data: data }
  end

  def chart id, data, chart
    page << { id: id, chart: chart, data: data }
  end

  def image id, data
    page << { id: id, data: data }
  end

  def to_html
    # Ruby emits non standard YAML that JavaScript can't parse. If you know how to make it to
    # output standard YAML, please fix it, as YAML is better format than JSON.
    #
    # require 'yaml'
    # Removing first line in YAML, ruby puts "--- !ruby/object:Page" there.
    # data = to_yaml.sub(/^---[^\n]+\n/, '')

    require 'json'
    # For strange reason `self.to_json` doesn't work, iterating over instance variables
    hash = {}
    instance_variables.each {|v| hash[v.to_s.delete("@")] = instance_variable_get(v) }
    data = hash.to_json

    @@standalone.sub('{type}', 'json').sub('{data}', data)
  end

  def save path
    File.write path, to_html
  end

  def publish url
    require 'net/http'
    require "uri"

    api_token = ENV['plot_api_token']
    raise "`plot_api_token` env variable is not set" if api_token.nil?
    resp = Net::HTTP.post URI(url), to_html, { api_token: api_token }
    raise "can't publish #{url}" unless resp.code == "200"
  end

  def self.unpublish url
    require 'net/http'
    require "uri"

    api_token = ENV['plot_api_token']
    raise "`plot_api_token` env variable is not set" if api_token.nil?

    # I don't know how to make HTTP DELETE request with Ruby, the code below doesn't work
    #
    # resp = Net::HTTP.delete URI(url, { api_token: api_token })
    # raise "can't delete #{url}" unless resp.code == "200"

    resp = `curl -s --request DELETE --header "api_token: #{api_token}" "#{url}"`
    raise "can't delete #{url}" unless /file deleted/ =~ resp
  end


  # standalone begin
  @@standalone = <<~HTML
  <!DOCTYPE html>
  <html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
  
    <title>pl0t</title>
  
    <link rel="icon" type="image/png" href="http://files.pl0t.com/view-1/favicon.ico">
  
    <!-- PL0T begin -->
    <script>
      window.env = {
        base_url: "http://files.pl0t.com/view-1"
      }
    </script>
    <link rel="stylesheet" href="http://files.pl0t.com/view-1/releases/2021-09-13-5cd3a7/bundle.css">
    <script defer src="http://files.pl0t.com/view-1/releases/2021-09-13-5cd3a7/bundle.js"></script>
    <!-- PL0T end -->
  </head>
  <body>
  
  <!-- PL0T data, type could be 'yaml', 'json' or 'md' -->
  <script id="data" type="{type}">
  {data}
  </script>
  
  
  <script>
    window.on_plot_loaded = function() {
      window.plot_api.run()
    }
  </script>
  
  </body>
  </html>
  HTML
  # standalone end
end